### ISSUE TITLE
java.net.SocketException : Too many open files while parsing DwC CSV

### ISSUE THREAD

**Issue by ansell** [AtlasOfLivingAustralia/biocache-store#153]
The following exception occurred while parsing a DwC CSV with a large number of image files:

```
2016-10-04 10:23:33,262 ERROR: [DataLoader] - Too many open files
java.net.SocketException: Too many open files
    at java.net.Socket.createImpl(Socket.java:447)
    at java.net.Socket.getImpl(Socket.java:510)
    at java.net.Socket.setSoTimeout(Socket.java:1105)
    at org.apache.http.conn.scheme.PlainSocketFactory.connectSocket(PlainSocketFactory.java:116)
    at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:178)
    at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:304)
    at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:610)
    at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:445)
    at org.apache.http.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:863)
    at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
    at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:106)
    at au.org.ala.biocache.util.HttpUtil$.postBody(HttpUtil.scala:23)
    at au.org.ala.biocache.load.RemoteMediaStore$.alreadyStored(MediaStore.scala:200)
    at au.org.ala.biocache.load.RemoteMediaStore$.save(MediaStore.scala:269)
    at au.org.ala.biocache.load.DataLoader$$anonfun$processMedia$1.apply(DataLoader.scala:316)
    at au.org.ala.biocache.load.DataLoader$$anonfun$processMedia$1.apply(DataLoader.scala:295)
    at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
    at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
    at au.org.ala.biocache.load.DataLoader$class.processMedia(DataLoader.scala:295)
    at au.org.ala.biocache.load.DwcCSVLoader.processMedia(DwcCSVLoader.scala:72)
    at au.org.ala.biocache.load.DataLoader$class.load(DataLoader.scala:257)
    at au.org.ala.biocache.load.DwcCSVLoader.load(DwcCSVLoader.scala:72)
    at au.org.ala.biocache.load.DwcCSVLoader.loadFile(DwcCSVLoader.scala:275)
    at au.org.ala.biocache.load.DwcCSVLoader$$anonfun$loadDirectory$1.apply(DwcCSVLoader.scala:125)
    at au.org.ala.biocache.load.DwcCSVLoader$$anonfun$loadDirectory$1.apply(DwcCSVLoader.scala:123)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
    at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:108)
    at au.org.ala.biocache.load.DwcCSVLoader.loadDirectory(DwcCSVLoader.scala:123)
    at au.org.ala.biocache.load.DwcCSVLoader$$anonfun$load$1.apply(DwcCSVLoader.scala:105)
    at au.org.ala.biocache.load.DwcCSVLoader$$anonfun$load$1.apply(DwcCSVLoader.scala:97)
    at scala.collection.immutable.List.foreach(List.scala:318)
    at au.org.ala.biocache.load.DwcCSVLoader.load(DwcCSVLoader.scala:97)
    at au.org.ala.biocache.load.Loader.load(Loader.scala:161)
    at au.org.ala.biocache.load.Loader$$anonfun$main$4.apply(Loader.scala:65)
    at au.org.ala.biocache.load.Loader$$anonfun$main$4.apply(Loader.scala:65)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
    at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:108)
    at au.org.ala.biocache.load.Loader$.main(Loader.scala:64)
    at au.org.ala.biocache.cmd.CMD2$.main(CMD2.scala:134)
    at au.org.ala.biocache.cmd.CMD2.main(CMD2.scala)
Exception in thread "main" java.net.SocketException: Too many open files
    at java.net.Socket.createImpl(Socket.java:447)
    at java.net.Socket.getImpl(Socket.java:510)
    at java.net.Socket.setSoTimeout(Socket.java:1105)
    at org.apache.http.conn.scheme.PlainSocketFactory.connectSocket(PlainSocketFactory.java:116)
    at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:178)
    at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:304)
    at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:610)
    at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:445)
    at org.apache.http.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:863)
    at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
    at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:106)
    at au.org.ala.biocache.util.HttpUtil$.postBody(HttpUtil.scala:23)
    at au.org.ala.biocache.load.RemoteMediaStore$.alreadyStored(MediaStore.scala:200)
    at au.org.ala.biocache.load.RemoteMediaStore$.save(MediaStore.scala:269)
    at au.org.ala.biocache.load.DataLoader$$anonfun$processMedia$1.apply(DataLoader.scala:316)
    at au.org.ala.biocache.load.DataLoader$$anonfun$processMedia$1.apply(DataLoader.scala:295)
    at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
    at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
    at au.org.ala.biocache.load.DataLoader$class.processMedia(DataLoader.scala:295)
    at au.org.ala.biocache.load.DwcCSVLoader.processMedia(DwcCSVLoader.scala:72)
    at au.org.ala.biocache.load.DataLoader$class.load(DataLoader.scala:257)
    at au.org.ala.biocache.load.DwcCSVLoader.load(DwcCSVLoader.scala:72)
    at au.org.ala.biocache.load.DwcCSVLoader.loadFile(DwcCSVLoader.scala:275)
    at au.org.ala.biocache.load.DwcCSVLoader$$anonfun$loadDirectory$1.apply(DwcCSVLoader.scala:125)
    at au.org.ala.biocache.load.DwcCSVLoader$$anonfun$loadDirectory$1.apply(DwcCSVLoader.scala:123)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
    at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:108)
    at au.org.ala.biocache.load.DwcCSVLoader.loadDirectory(DwcCSVLoader.scala:123)
    at au.org.ala.biocache.load.DwcCSVLoader$$anonfun$load$1.apply(DwcCSVLoader.scala:105)
    at au.org.ala.biocache.load.DwcCSVLoader$$anonfun$load$1.apply(DwcCSVLoader.scala:97)
    at scala.collection.immutable.List.foreach(List.scala:318)
    at au.org.ala.biocache.load.DwcCSVLoader.load(DwcCSVLoader.scala:97)
    at au.org.ala.biocache.load.Loader.load(Loader.scala:161)
    at au.org.ala.biocache.load.Loader$$anonfun$main$4.apply(Loader.scala:65)
    at au.org.ala.biocache.load.Loader$$anonfun$main$4.apply(Loader.scala:65)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
    at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:108)
    at au.org.ala.biocache.load.Loader$.main(Loader.scala:64)
    at au.org.ala.biocache.cmd.CMD2$.main(CMD2.scala:134)
    at au.org.ala.biocache.cmd.CMD2.main(CMD2.scala)
```

It isn't clear if the error is related directly to Socket's or more generally to File Descriptors. In both cases it may be related to not actively closing File or Socket references after use.

The stacktrace appears to be serial and synchronous (not parallel or asynchronous) and Jenkins is only configured with 2(?) workers.

**Comment by ansell** [comment]
The method in question here, HttpUtil.postBody, has a resource leak in cases where exceptions are thrown or the stream is not completely read. It needs to have releaseConnection called in a finally block to fix the immediate resource leak, but there are also other instances where HttpClient/connections are not closed/released correctly that will also contribute to this and other resource leaks.

**Comment by ansell** [comment]
This error should also have been fixed by #156, and difficult to test as it relies on services being flaky. Closing but can reopen in future if it occurs again.

**Comment by ansell** [comment]
Refs https://github.com/AtlasOfLivingAustralia/maintenance/issues/35
